src/widget/switch.rs

```diff
@@ -0,0 +1,242 @@
+// Copyright 2019 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A toggle switch widget.
+
+use crate::kurbo::{Circle, Point, Rect, RoundedRect, Size};
+use crate::piet::{FontBuilder, Text, TextLayout, TextLayoutBuilder};
+use crate::piet::{LinearGradient, RenderContext, UnitPoint};
+use crate::theme;
+use crate::widget::Align;
+use crate::{
+    BaseState, BoxConstraints, Env, Event, EventCtx, LayoutCtx, PaintCtx, UpdateCtx, Widget,
+};
+
+#[derive(Debug, Clone)]
+pub struct Switch;
+
+impl Switch {
+    pub fn new() -> impl Widget<bool> {
+        Align::vertical(UnitPoint::CENTER, SwitchRaw::default())
+    }
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct SwitchRaw {
+    knob_pos: Point,
+    knob_hovered: bool,
+}
+
+impl SwitchRaw {
+    fn knob_hit_test(&self, knob_width: f64, mouse_pos: Point) -> bool {
+        let knob_circle = Circle::new(self.knob_pos, knob_width / 2.);
+        if mouse_pos.distance(knob_circle.center) < knob_circle.radius {
```

Comment:
the less obvious but maybe canonical way to do this might be `if knob_circle.winding(mouse_pos) > 0 { .. ` but personally I think that the `Shape` trait should just have a `contains(pt: Point)` type method that calls `winding` under the hood; I think it would be a much simpler API to discover and to use, in many cases.